File: src/lib/batteryService.ts
import { ObjectId, WithId } from 'mongodb';
import { BatteryConfig } from '@/types';
import { getDb } from './db';

const COLLECTION_NAME = 'baterias';

type BatteryDocument = Omit<BatteryConfig, '_id' | 'createdAt' | 'updatedAt'> & {
  createdAt: Date;
  updatedAt: Date;
};

function mapBattery(doc: WithId<BatteryDocument>): BatteryConfig {
  return {
    _id: doc._id.toHexString(),
    manufacturer: doc.manufacturer,
    model: doc.model,
    capacityKwh: doc.capacityKwh,
    quantity: doc.quantity,
    createdAt: doc.createdAt.toISOString(),
    updatedAt: doc.updatedAt.toISOString(),
  };
}

function ensurePositive(value: unknown, field: string): number {
  const num = Number(value);
  if (!Number.isFinite(num) || num <= 0) {
    throw new Error(`El campo ${field} debe ser un número mayor que cero.`);
  }
  return num;
}

export async function listBatteries(): Promise<BatteryConfig[]> {
  const db = await getDb();
  const collection = db.collection<BatteryDocument>(COLLECTION_NAME);
  const docs = await collection.find().sort({ updatedAt: -1 }).toArray();
  return docs.map(mapBattery);
}

export async function getBatteryById(id: string): Promise<BatteryConfig | null> {
  const db = await getDb();
  const collection = db.collection<BatteryDocument>(COLLECTION_NAME);
  const _id = new ObjectId(id);
  const doc = await collection.findOne({ _id });
  return doc ? mapBattery(doc) : null;
}

export async function createBattery(payload: Partial<BatteryConfig>): Promise<BatteryConfig> {
  if (!payload.manufacturer?.trim()) {
    throw new Error('El campo manufacturer es obligatorio.');
  }

  const now = new Date();
  const document: BatteryDocument = {
    manufacturer: payload.manufacturer.trim(),
    model: payload.model,
    capacityKwh: ensurePositive(payload.capacityKwh, 'capacityKwh'),
    quantity: Math.trunc(ensurePositive(payload.quantity, 'quantity')),
    createdAt: now,
    updatedAt: now,
  };

  const db = await getDb();
  const collection = db.collection<BatteryDocument>(COLLECTION_NAME);
  const result = await collection.insertOne(document);

  return {
    ...payload,
    _id: result.insertedId.toHexString(),
    manufacturer: document.manufacturer,
    model: document.model,
    capacityKwh: document.capacityKwh,
    quantity: document.quantity,
    createdAt: document.createdAt.toISOString(),
    updatedAt: document.updatedAt.toISOString(),
  };
}

export async function updateBattery(
  id: string,
  payload: Partial<BatteryConfig>
): Promise<BatteryConfig | null> {
  const db = await getDb();
  const collection = db.collection<BatteryDocument>(COLLECTION_NAME);
  const _id = new ObjectId(id);

  const existing = await collection.findOne({ _id });
  if (!existing) {
    return null;
  }

  const update: Partial<BatteryDocument> = {};

  if (payload.manufacturer !== undefined) {
    if (!payload.manufacturer.trim()) {
      throw new Error('El campo manufacturer es obligatorio.');
    }
    update.manufacturer = payload.manufacturer.trim();
  }
  if (payload.model !== undefined) update.model = payload.model;
  if (payload.capacityKwh !== undefined) {
    update.capacityKwh = ensurePositive(payload.capacityKwh, 'capacityKwh');
  }
  if (payload.quantity !== undefined) {
    update.quantity = Math.trunc(ensurePositive(payload.quantity, 'quantity'));
  }

  update.updatedAt = new Date();

  await collection.updateOne({ _id }, { $set: update });
  const refreshed = await collection.findOne({ _id });

  return refreshed ? mapBattery(refreshed) : null;
}

export async function deleteBattery(id: string): Promise<boolean> {
  const db = await getDb();
  const collection = db.collection<BatteryDocument>(COLLECTION_NAME);
  const _id = new ObjectId(id);
  const result = await collection.deleteOne({ _id });
  return result.deletedCount === 1;
}

File: backend/app/services/battery_service.py
"""
Battery bank CRUD helpers backed by MongoDB.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from bson import ObjectId
from pymongo.collection import Collection

from app.database import get_database

COLLECTION_NAME = "baterias"


def _collection() -> Collection:
    return get_database()[COLLECTION_NAME]


def _ensure_positive(value: Any, field: str) -> float:
    if value is None:
        raise ValueError(f"El campo {field} es obligatorio.")
    number = float(value)
    if number <= 0:
        raise ValueError(f"El campo {field} debe ser un número mayor que cero.")
    return number


def _ensure_text(value: Any, field: str) -> str:
    if value is None:
        raise ValueError(f"El campo {field} es obligatorio.")
    text = str(value).strip()
    if not text:
        raise ValueError(f"El campo {field} es obligatorio.")
    return text


def _object_id(battery_id: str) -> ObjectId:
    try:
        return ObjectId(battery_id)
    except Exception as exc:  # pragma: no cover
        raise ValueError("Identificador de batería inválido.") from exc


def _map_battery(doc: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "_id": str(doc["_id"]),
        "manufacturer": doc.get("manufacturer"),
        "model": doc.get("model"),
        "capacityKwh": doc.get("capacityKwh"),
        "quantity": doc.get("quantity"),
        "createdAt": doc.get("createdAt").isoformat() if doc.get("createdAt") else None,
        "updatedAt": doc.get("updatedAt").isoformat() if doc.get("updatedAt") else None,
    }


def list_batteries() -> List[Dict[str, Any]]:
    cursor = _collection().find().sort("updatedAt", -1)
    return [_map_battery(doc) for doc in cursor]


def get_battery(battery_id: str) -> Optional[Dict[str, Any]]:
    doc = _collection().find_one({"_id": _object_id(battery_id)})
    return _map_battery(doc) if doc else None


def create_battery(payload: Dict[str, Any]) -> Dict[str, Any]:
    manufacturer = _ensure_text(payload.get("manufacturer"), "manufacturer")

    now = datetime.utcnow()
    document: Dict[str, Any] = {
        "manufacturer": manufacturer,
        "model": payload.get("model"),
        "capacityKwh": _ensure_positive(payload.get("capacityKwh"), "capacityKwh"),
        "quantity": int(_ensure_positive(payload.get("quantity"), "quantity")),
        "createdAt": now,
        "updatedAt": now,
    }
    result = _collection().insert_one(document)
    document["_id"] = result.inserted_id
    return _map_battery(document)


def update_battery(battery_id: str, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    update: Dict[str, Any] = {}

    if "manufacturer" in payload:
        update["manufacturer"] = _ensure_text(payload.get("manufacturer"), "manufacturer")
    if "model" in payload:
        update["model"] = payload.get("model")

    numeric_fields = {
        "capacityKwh": _ensure_positive,
        "quantity": lambda value, field: int(_ensure_positive(value, field)),
    }

    for field, validator in numeric_fields.items():
        if field in payload and payload[field] is not None:
            update[field] = validator(payload[field], field)

    if not update:
        return get_battery(battery_id)

    update["updatedAt"] = datetime.utcnow()
    result = _collection().find_one_and_update(
        {"_id": _object_id(battery_id)},
        {"$set": update},
        return_document=True,
    )
    return _map_battery(result) if result else None


def delete_battery(battery_id: str) -> bool:
    result = _collection().delete_one({"_id": _object_id(battery_id)})
    return result.deleted_count == 1
